### ISSUE TITLE
Tricko - Rewards removed from `extraRewards` array may become stuck in the contract.

### ISSUE THREAD

**Issue by sherlock-admin4** [sherlock-audit/2024-03-zivoe-judging#639]
Tricko

medium

# Rewards removed from `extraRewards` array may become stuck in the contract.

## Summary
As a consequence of only checking the current `extraRewards` during the `claimRewards()` it is not possible to transfer rewards received previously but that have been removed from the `extraRewards` array.

## Vulnerability Detail
Both the `OCY_Convex_A` and `OCY_Convex_C` contracts are lockers for depositing funds and interacting with pools on both [Curve](https://curve.fi) and [Convex](https://www.convexfinance.com/). These contracts facilitate the accumulation and transfer of `CRX` and `CRV` rewards over time for positions within Convex, with the potential for additional rewards.

To forward those rewards back to the `OCT_YDL`, both contracts (`OCY_Convex_A` and `OCY_Convex_C`) have the `claimRewards()` function.

```solidity
    function claimRewards(bool extra) public nonReentrant {
        IBaseRewardPool_OCY_Convex_A(convexRewards).getReward();


        // Native Reward (CRV)
        uint256 rewardsCRV = IERC20(CRV).balanceOf(address(this));
        uint256 rewardsCVX = IERC20(CVX).balanceOf(address(this));
        if (rewardsCRV > 0) { IERC20(CRV).safeTransfer(OCT_YDL, rewardsCRV); }
        if (rewardsCVX > 0) { IERC20(CVX).safeTransfer(OCT_YDL, rewardsCVX); }


        // Extra Rewards
        if (extra) {
            uint256 extraRewardsLength = IBaseRewardPool_OCY_Convex_A(convexRewards).extraRewardsLength();
            for (uint256 i = 0; i < extraRewardsLength; i++) {
                address rewardContract = IBaseRewardPool_OCY_Convex_A(convexRewards).extraRewards(i);
                uint256 rewardAmount = IBaseRewardPool_OCY_Convex_A(rewardContract).rewardToken().balanceOf(
                    address(this)
                );
                if (rewardAmount > 0) { IERC20(rewardContract).safeTransfer(OCT_YDL, rewardAmount); }
            }
        }
    }
```

As we can see from the code, if `extra` function argument is `true` then the code iterates over the extra rewards and transfer those if available to the `OCT_YDL` contract. However, there is a limitation: the `claimRewards()` function only considers the rewards present in the `extraRewards` array at the time of its execution. If rewards were received previously but have since been removed from this array, they cannot be forwarded to `OCT_YDL`.

Consider the following scenario:

- Time T . Reward Y added.
extraRewards = [X, Y]
Contract balances = {x:100, y:0}

- Time T + 3 days.
extraRewards = [X, Y]
Contract balances = {x:120, y:20}

- Time T + 6 days. Reward Y removed
extraRewards = [X]
Contract balances = {x:140, y:40}

- Time T + 9 days. `claimRewards(true)` is called.
extraRewards = [X]
Contract balances = {x:0, y:40}

Even though we have nonzero Y balance, we cannot transfer it because at the time of  `claimRewards()` call, the Y token is not in the `extraRewards` array.

To summarize. Due to checking the current state of `extraRewards` during `claimRewards()`, rewards previously received but subsequently removed from the array cannot be forwarded to `OCT_YDL`.

## Impact
Convex's extra rewards that have been added and removed before the next call to `claimRewards(true)` will become trapped in the contract.

## Code Snippet
https://github.com/sherlock-audit/2024-03-zivoe/blob/main/zivoe-core-foundry/src/lockers/OCY/OCY_Convex_A.sol#L246-L266

https://github.com/sherlock-audit/2024-03-zivoe/blob/main/zivoe-core-foundry/src/lockers/OCY/OCY_Convex_C.sol#L200-L219

## Tool used
Manual Review

## Recommendation
Consider  implementing an additional function which allows for the transfer of arbitrary tokens to `OCT_YDL`

Duplicate of #506
